File: python/VCFSelfFillable.py
from __future__ import annotations

# coding: utf-8
# VCFFillable.py

from itertools import *
from collections import defaultdict

from typing import Optional

from VCF import *
from VCFGeno import VCFGenoBase, VCFGeno
from GenoRecord import GenoRecord
from VCFSelfHetero import VCFSelfHetero
from VCFImpSelfRecord import SelfFillType, VCFImpSelfRecord
from VCFSelfFillableRecord import VCFSelfFillableRecord
from VCFHeteroHomo import *
from SelfGroups import SelfGroups
from SelfRecordSet import SelfRecordSet
from Genotype import Genotype
from option import *
from common import *


#################### VCFSelfFillable ####################

class VCFSelfFillable(VCFGenoBase):
	def __init__(self, samples: list[str],
						records: list[VCFSelfFillableRecord],
						vcf: VCFSmall) -> None:
		VCFGenoBase.__init__(self, samples, vcf)
		self.records: list[VCFSelfFillableRecord] = records
	
	def __len__(self) -> int:
		return len(self.records)
	
	def get_record(self, i: int) -> GenoRecord:
		return self.records[i]
	
	def modify(self) -> None:
		# FillTypeでrecordを分ける
		groups = SelfGroups.create(self.records)
		for record_set in groups.generate_record_sets():
			record_set.determine_parents_phasing()
	
	@staticmethod
	def fill(vcfs: list[VCFSelfHetero],
					records: list[VCFImpSelfRecord]) -> VCFSelfFillable:
		merged_records = VCFSelfFillable.merge_records(vcfs, records)
		vcf = VCFSelfFillable(vcfs[0].samples, merged_records, vcfs[0].vcf)
		vcf.modify()
		return vcf
	
	@staticmethod
	def merge_records(vcfs: list[VCFSelfHetero], records: list[VCFImpSelfRecord]
											) -> list[VCFSelfFillableRecord]:
		all_records = [ record for record in chain(
							(r for vcf in vcfs for r in vcf.records), records) ]
		all_records.sort(key=lambda record: record.index)
		probs = VCFSelfFillable.calc_probs(all_records, vcfs[0])
		new_records = [ VCFSelfFillableRecord.convert(r, p)
									for r, p in zip(all_records, probs) ]
		return new_records
	
	@staticmethod
	def calc_probs(records: list[VCFImpSelfRecord],
					vcf: VCFGenoBase) -> list[list[tuple[float, float, float]]]:
		orig_vcf = vcf.vcf
		cols = orig_vcf.extract_columns(vcf.samples)
		prob_table = []
		for record, orig_record in zip(records, orig_vcf.records):
			probs = orig_record.parse_PL(record.geno, cols)
			prob_table.append(probs)
		return prob_table

File: src/VCFSelfFillable.cpp
#include <limits>
#include <memory>
#include <cassert>
#include "../include/VCFSelfFillable.h"
#include "../include/VCFSelfHetero.h"
#include "../include/SelfGroups.h"
#include "../include/option.h"
#include "../include/common.h"

using namespace std;


//////////////////// VCFSelfFillable ////////////////////

VCFSelfFillable::VCFSelfFillable(const STRVEC& s,
							const std::vector<VCFSelfFillableRecord *>& rs,
							const VCFSmall *vcf) :
										VCFGenoBase(s, vcf), records(rs) { }

VCFSelfFillable::~VCFSelfFillable() {
	for(auto p = records.begin(); p != records.end(); ++p)
		delete *p;
}

void VCFSelfFillable::modify() {
	const auto	*groups = SelfGroups::create(records);
	const auto	record_sets = groups->create_record_sets();
	for(auto p = record_sets.begin(); p != record_sets.end(); ++p) {
		(*p)->determine_parent_phasing();
	}
	
	delete groups;
	Common::delete_all(record_sets);
}

VCFSelfFillable *VCFSelfFillable::fill(const vector<VCFSelfHetero *>& vcfs,
									const vector<VCFImpSelfRecord *>& records) {
	vector<VCFSelfFillableRecord *>	merged_records
							 = VCFSelfFillable::merge_records(vcfs, records);
	VCFSelfFillable	*vcf = new VCFSelfFillable(vcfs.front()->get_samples(),
												merged_records,
												vcfs.front()->get_ref_vcf());
	vcf->modify();
	return vcf;
}

vector<VCFSelfFillableRecord *> VCFSelfFillable::merge_records(
									const vector<VCFSelfHetero *>& vcfs,
									const vector<VCFImpSelfRecord *>& records) {
	vector<VCFImpSelfRecord *>	all_records;
	for(auto p = vcfs.begin(); p != vcfs.end(); ++p) {
		auto	*vcf = *p;
		auto&	rs = vcf->get_records();
		all_records.insert(all_records.end(), rs.begin(), rs.end());
	}
	all_records.insert(all_records.end(), records.begin(), records.end());
	std::sort(all_records.begin(), all_records.end(),
				[](const VCFImpSelfRecord *r1, const VCFImpSelfRecord *r2) {
					return r1->get_index() < r2->get_index(); });
	const auto&	probs = VCFSelfFillable::calc_probs(all_records, vcfs[0]);
	vector<VCFSelfFillableRecord *>	new_records;
	for(size_t i = 0; i < all_records.size(); ++i) {
		auto	*r = VCFSelfFillableRecord::convert(all_records[i], probs[i]);
		new_records.push_back(r);
	}
	return new_records;
}

vector<vector<VCFRecord::Probs>> VCFSelfFillable::calc_probs(
									const vector<VCFImpSelfRecord *>& records,
									const VCFGenoBase *vcf) {
	const VCFSmall	*orig_vcf = vcf->get_ref_vcf();
	const auto	cols = orig_vcf->extract_columns(vcf->get_samples());
	vector<vector<VCFRecord::Probs>>	prob_table;
	for(size_t i = 0; i < records.size(); ++i) {
		const auto&	geno = records[i]->get_geno();
		const auto	probs = orig_vcf->get_record(i)->parse_PL(geno, cols);
		prob_table.push_back(probs);
	}
	return prob_table;
}
